xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java

```diff
@@ -0,0 +1,471 @@
+  }
+
+  private void afterAcceptAddresses() {
+    for (Subchannel subchannel : getSubchannels()) {
+      WrrSubchannel weightedSubchannel = (WrrSubchannel) subchannel;
+      weightedSubchannel.setConfig(config);
+      if (config.enableOobLoadReport) {
+        OrcaOobUtil.setListener(weightedSubchannel, weightedSubchannel.oobListener,
+                OrcaOobUtil.OrcaReportingConfig.newBuilder()
+                        .setReportInterval(config.oobReportingPeriodNanos, TimeUnit.NANOSECONDS)
+                        .build());
+      } else {
+        OrcaOobUtil.setListener(weightedSubchannel, null, null);
+      }
+    }
+  }
+
+  @Override
+  public void shutdown() {
+    if (weightUpdateTimer != null) {
+      weightUpdateTimer.cancel();
+    }
+    super.shutdown();
+  }
+
+  private static final class WrrHelper extends ForwardingLoadBalancerHelper {
+    private final Helper delegate;
+    private final TimeProvider timeProvider;
+
+    WrrHelper(Helper helper, TimeProvider timeProvider) {
+      this.delegate = helper;
+      this.timeProvider = timeProvider;
+    }
+
+    @Override
+    protected Helper delegate() {
+      return delegate;
+    }
+
+    @Override
+    public Subchannel createSubchannel(CreateSubchannelArgs args) {
+      return new WrrSubchannel(delegate().createSubchannel(args), timeProvider);
+    }
+  }
+
+  @VisibleForTesting
+  static final class WrrSubchannel extends ForwardingSubchannel {
+    private final Subchannel delegate;
+    private final TimeProvider timeProvider;
+    private final OrcaOobReportListener oobListener = this::onLoadReport;
+    private final OrcaPerRequestReportListener perRpcListener = this::onLoadReport;
+    private volatile long lastUpdated;
+    private volatile long nonEmptySince;
+    private volatile double weight;
+    private volatile WeightedRoundRobinLoadBalancerConfig config;
+
+    WrrSubchannel(Subchannel delegate, TimeProvider timeProvider) {
+      this.delegate = checkNotNull(delegate, "delegate");
+      this.timeProvider = checkNotNull(timeProvider, "timeProvider");
+    }
+
+    private void setConfig(WeightedRoundRobinLoadBalancerConfig config) {
+      this.config = config;
+    }
+
+    @VisibleForTesting
+    void onLoadReport(MetricReport report) {
+      double newWeight = report.getCpuUtilization() == 0 ? 0 :
+              report.getQps() / report.getCpuUtilization();
+      if (newWeight == 0) {
+        return;
+      }
+      if (nonEmptySince == Integer.MAX_VALUE) {
+        nonEmptySince = timeProvider.currentTimeNanos();
+      }
+      lastUpdated = timeProvider.currentTimeNanos();
+      weight = newWeight;
+    }
+
+    @Override
+    public void start(SubchannelStateListener listener) {
+      delegate().start(new SubchannelStateListener() {
+        @Override
+        public void onSubchannelState(ConnectivityStateInfo newState) {
+          if (newState.getState().equals(ConnectivityState.READY)) {
+            nonEmptySince = Integer.MAX_VALUE;
```

Comment:
oh I think I made the `WrrSubchannel` static because static wrrHelper (need to be static because constructed before super class of `WeightedRoundRobinLoadBalancer` is constructed) uses it.

Reply:
The current inheritance approach is mostly still clean, but this does cause inconvenience. (static and non-static workarounds, super class construction order). I am thinking using `ForwardingLoadBalancer` with a subchannel creation factory, potentially to simplify in the future.

I can understand if `timeProvider.currentTimeNanos()` reaches long.MAX_VALUE or Integer.MAX_VALUE,  there is a problem about collapsing with unset time logic. But when the time value wraps, the algorithm is just not working in a even more ugly situation. Preventing the collapse may become minor. How to prevent that happen?

Reply:
(As pointed out in "offline" discussions, we should be using Deadline.Ticker as it is powered by nanoTime(), not currentTimeMillis(). Same as https://github.com/grpc/grpc-java/issues/9048 )

It looks like time wrapping properly doesn't cause an issue. `nanoTime()` is only useful for durations, and so you [always subtract one time from another time](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/System.html#nanoTime()).

But yeah, the algorithm would be broken by using UNSET_TIME in the past. We can use nanoTime() + 100 years or nanoTime() + Long.MAX_VALUE instead.